### DISCUSSION TITLE
Parsing `&[u8]` as e.g. `&[u16]`

### DISCUSSION THREAD

**Discussion by DJDuque** [winnow-rs/winnow#586] [Ideas]
I was wondering if it would make sense to have some binary parsers that do something like (and similar for other integer types):

```
fn parser(input: &mut &[u8]) -> PResult<Cow<'_, [u16]>, Error> {
    ...
}
```

I am thinking of use cases where you want to parse a binary data packet like:
```
struct Packet {
    header: u8,
    version: u32,
    data: Vec<u16>,
}
```

We could, instead have a:
```
struct PacketView<'a> {
    header: u8,
    version: u32,
    data: Cow<'a, [u16]>,
}
```
And get rid of the vector allocation if it is not needed. If target endianness matches data endianness and alignment allow, we get a slice of `u16`, otherwise we just allocate the vector.

Does something like this make sense in winnow, and is it worth it?

**Comment by epage** [comment]
> Does something like this make sense in winnow, and is it worth it?

While we have [guidelines on parsers](https://github.com/winnow-rs/winnow/blob/main/CONTRIBUTING.md#parsers), we haven't figured out the right line for what should be built-in.

Everyone has a different use case which will emphasize different needs.  For example, I do a lot of text parsing and only use the higher level capabilities for prototyping.  This makes it more difficult to evaluate.

It doesn't help that there is little lost in having these things be external.

Could you provide examples of formats that require slice conversions like that?  Looking to better understand how often it comes up and what data types the conversions happen for.

**Reply by DJDuque** [reply]
> Could you provide examples of formats that require slice conversions like that? Looking to better understand how often it comes up and what data types the conversions happen for.

I work mostly with custom formats from detectors in physics experiments. Here, it is very common to have data packets from e.g. an ADC with digitized samples of a signal/waveform. I don't know about popular formats, but I assume this would also be very common dealing with audio (?) e.g. [WAV](https://en.wikipedia.org/wiki/WAV) format has a `wave-data` chunk.

Data types would usually be `i16`, `i32`; but it really applies for all integer types. Actually, could this be generalized to `T`, where `T` is `repr(C)`?

**Reply by epage** [reply]
> Actually, could this be generalized to T, where T is repr(C)?

Whether we'd need to or not is partly why I was asking.  It might get messy dealing with `Cow` but it is likely doable.  I'd recommend giving this a try in your code base, see how it works, and then we can evaluate.
